File: examples/js-worker/worker.js
/**
 * GoQueue JavaScript Worker
 *
 * 适用环境：Node.js（需安装 ws 包）
 * 安装依赖：npm install ws
 * 启动方式：node worker.js
 *
 * 环境变量：
 *   GOQUEUE_SERVER   WebSocket 服务端地址，默认 ws://localhost:8080/ws/worker
 *   GOQUEUE_QUEUE    监听的队列名，默认 default
 *   GOQUEUE_API_KEY  API Key（对应服务端 API_KEY 环境变量）
 *
 * 新特性（v4）：
 *   - tags：任务可携带标签，handler 第二参数接收 tags 数组
 *   - batch catch/finally：批次失败/完成回调
 *   - queue pause/resume：队列可暂停，暂停期间任务不派发
 */

'use strict';

const WebSocket = require('ws');

// ─── 配置 ────────────────────────────────────────────────────────────────────
const SERVER_URL      = process.env.GOQUEUE_SERVER  || 'ws://localhost:8080/ws/worker';
const QUEUE           = process.env.GOQUEUE_QUEUE   || 'default';
const API_KEY         = process.env.GOQUEUE_API_KEY || '';
const CONCURRENCY     = parseInt(process.env.GOQUEUE_CONCURRENCY || '4', 10);
const PING_INTERVAL   = 20_000;   // 心跳间隔（ms）
const RECONNECT_DELAY = 3_000;    // 断线重连间隔（ms）

// ─── Job Handlers ─────────────────────────────────────────────────────────────
// handler 签名：async (data, tags) => string
// data: 解析后的 payload 对象
// tags: 任务标签数组（v4 新增），例如 ["urgent", "notify"]

const handlers = {

  send_email: async (data, tags) => {
    console.log(`[send_email] to=${data.to} subject=${data.subject} tags=${JSON.stringify(tags)}`);
    await sleep(300);
    return `Email sent to ${data.to}`;
  },

  generate_report: async (data, tags) => {
    console.log(`[generate_report] name=${data.name} tags=${JSON.stringify(tags)}`);
    await sleep(800);
    return `Report "${data.name}" generated`;
  },

  resize_image: async (data, tags) => {
    console.log(`[resize_image] url=${data.url} size=${data.width}x${data.height} tags=${JSON.stringify(tags)}`);
    await sleep(500);
    return `Image ${data.url} resized to ${data.width}x${data.height}`;
  },

  data_sync: async (data, tags) => {
    console.log(`[data_sync] ${data.source} → ${data.target} tags=${JSON.stringify(tags)}`);
    await sleep(600);
    return `Synced ${data.source} → ${data.target}`;
  },

  // v4: 按 tags 路由处理示例
  tag_task: async (data, tags) => {
    console.log(`[tag_task] message=${data.message} tags=${JSON.stringify(tags)}`);
    if (tags.includes('dry-run')) {
      console.log('[tag_task] dry-run 模式，跳过实际操作');
      return 'dry-run: skipped';
    }
    if (tags.includes('urgent')) {
      console.log('[tag_task] 紧急任务，优先处理');
    }
    await sleep(200);
    return `tag_task done: ${data.message} (tags=${JSON.stringify(tags)})`;
  },

  // v4: batch 回调示例（then_job / catch_job / finally_job 均可复用）
  batch_callback: async (data, tags) => {
    console.log(`[batch_callback] batch_id=${data.batch_id} status=${data.status} tags=${JSON.stringify(tags)}`);
    // 在这里可以发送通知、更新数据库等
    return `batch ${data.batch_id} callback handled`;
  },

  on_success: async (data, tags) => {
    console.log(`[on_success] batch_id=${data.batch_id} tags=${JSON.stringify(tags)}`);
    return `batch ${data.batch_id} succeeded`;
  },

  on_failure: async (data, tags) => {
    console.log(`[on_failure] batch_id=${data.batch_id} tags=${JSON.stringify(tags)}`);
    return `batch ${data.batch_id} failed`;
  },

  on_finally: async (data, tags) => {
    console.log(`[on_finally] batch_id=${data.batch_id} tags=${JSON.stringify(tags)}`);
    return `batch ${data.batch_id} finished`;
  },
};

// ─── Worker 核心 ──────────────────────────────────────────────────────────────
let ws        = null;
let pingTimer = null;
let stopped   = false;
let running   = 0;   // 当前并发任务数

function connect() {
  if (stopped) return;

  const url = `${SERVER_URL}?queue=${QUEUE}`;
  const wsOptions = API_KEY ? { headers: { Authorization: `Bearer ${API_KEY}` } } : {};

  console.log(`[Worker] Connecting to ${url} ...`);
  ws = new WebSocket(url, wsOptions);

  ws.on('open', () => {
    console.log('[Worker] Connected ✓');
    pingTimer = setInterval(() => {
      if (ws.readyState === WebSocket.OPEN) ws.ping();
    }, PING_INTERVAL);
  });

  ws.on('message', (raw) => {
    let msg;
    try { msg = JSON.parse(raw); } catch { return; }

    switch (msg.type) {
      case 'connected':
        console.log(`[Worker] Server: ${msg.message}`);
        break;
      case 'job':
        if (running >= CONCURRENCY) {
          // 超出并发限制，拒绝（服务端会重新入队）
          sendResult(msg.job_id, false, null, 'worker busy');
          return;
        }
        running++;
        handleJob(msg).finally(() => running--);
        break;
      case 'ack':
        console.log(`[Worker] ACK: ${msg.message}`);
        break;
      default:
        console.log(`[Worker] Unknown type: ${msg.type}`);
    }
  });

  ws.on('close', () => {
    clearPing();
    if (!stopped) {
      console.log(`[Worker] Disconnected, reconnecting in ${RECONNECT_DELAY}ms...`);
      setTimeout(connect, RECONNECT_DELAY + Math.random() * 1000);
    }
  });

  ws.on('error', (err) => {
    console.error(`[Worker] WS error: ${err.message}`);
  });
}

async function handleJob(msg) {
  const { job_id, job_type, queue, payload, tags = [] } = msg;
  console.log(`[Job #${job_id}] type=${job_type} queue=${queue} tags=${JSON.stringify(tags)}`);

  let data;
  try { data = JSON.parse(payload); } catch (e) {
    sendResult(job_id, false, null, `payload parse error: ${e.message}`);
    return;
  }

  const handler = handlers[job_type];
  if (!handler) {
    console.warn(`[Job #${job_id}] No handler for job_type=${job_type}`);
    sendResult(job_id, false, null, `no handler for job_type: ${job_type}`);
    return;
  }

  try {
    const result = await handler(data, tags);
    console.log(`[Job #${job_id}] OK: ${result}`);
    sendResult(job_id, true, result, null);
  } catch (err) {
    console.error(`[Job #${job_id}] FAILED: ${err.message}`);
    sendResult(job_id, false, null, err.message);
  }
}

function sendResult(jobId, success, log, error) {
  if (!ws || ws.readyState !== WebSocket.OPEN) return;
  const msg = { type: 'result', job_id: jobId, success };
  if (success) msg.log   = log;
  else         msg.error = error;
  ws.send(JSON.stringify(msg));
}

function clearPing() {
  if (pingTimer) { clearInterval(pingTimer); pingTimer = null; }
}

function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

// ─── 优雅退出 ─────────────────────────────────────────────────────────────────
process.on('SIGINT',  () => { stopped = true; clearPing(); ws && ws.close(); process.exit(0); });
process.on('SIGTERM', () => { stopped = true; clearPing(); ws && ws.close(); process.exit(0); });

// ─── 启动 ─────────────────────────────────────────────────────────────────────
console.log(`[Worker] Queue=${QUEUE} Concurrency=${CONCURRENCY} Handlers=${Object.keys(handlers).join(', ')}`);
connect();

File: examples/python-worker/worker.py
#!/usr/bin/env python3
"""
GoQueue Python Worker 示例
依赖：pip install websocket-client
启动：python worker.py

环境变量：
  GOQUEUE_SERVER        WebSocket 服务端地址，默认 ws://localhost:8080/ws/worker
  GOQUEUE_QUEUE         监听的队列名，默认 default
  GOQUEUE_API_KEY       API Key（对应服务端 API_KEY 环境变量）
  GOQUEUE_CONCURRENCY   最大并发任务数，默认 4

新特性（v4）：
  - tags：任务可携带标签，handler 第二参数接收 tags 列表
  - batch catch/finally：批次失败/完成回调
  - queue pause/resume：队列可暂停，暂停期间任务不派发
"""

import json
import logging
import os
import threading
import time
from typing import Any, Callable, Dict, List, Optional

import websocket

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(message)s",
)
log = logging.getLogger("goqueue-worker")

# ─── 配置 ─────────────────────────────────────────────────────────────────────
SERVER_URL    = os.getenv("GOQUEUE_SERVER",      "ws://localhost:8080/ws/worker")
QUEUE         = os.getenv("GOQUEUE_QUEUE",       "default")
API_KEY       = os.getenv("GOQUEUE_API_KEY",     "")
CONCURRENCY   = int(os.getenv("GOQUEUE_CONCURRENCY", "4"))
PING_INTERVAL = 20      # 心跳间隔（秒）
RECONNECT_DELAY = 3.0   # 断线重连间隔（秒）

# ─── Handler 类型 ──────────────────────────────────────────────────────────────
# handler 签名：(data: dict, tags: list[str]) -> str
# data: 解析后的 payload 字典
# tags: 任务标签列表（v4 新增），例如 ["urgent", "notify"]
HandlerFunc = Callable[[Dict[str, Any], List[str]], str]

# ─── GoQueueWorker ─────────────────────────────────────────────────────────────
class GoQueueWorker:
    def __init__(self):
        self._handlers: Dict[str, HandlerFunc] = {}
        self._ws: Optional[websocket.WebSocketApp] = None
        self._stopped = False
        self._sem = threading.Semaphore(CONCURRENCY)

    def register(self, job_type: str, handler: HandlerFunc):
        """注册 job_type 对应的处理函数"""
        self._handlers[job_type] = handler
        log.info(f"Registered handler: {job_type}")

    def run(self):
        """启动 Worker（阻塞，自动重连）"""
        while not self._stopped:
            self._connect()
            if not self._stopped:
                log.info(f"Reconnecting in {RECONNECT_DELAY}s...")
                time.sleep(RECONNECT_DELAY)

    def stop(self):
        self._stopped = True
        if self._ws:
            self._ws.close()

    def _connect(self):
        url = f"{SERVER_URL}?queue={QUEUE}"
        headers = {}
        if API_KEY:
            headers["Authorization"] = f"Bearer {API_KEY}"

        log.info(f"Connecting to {url} ...")

        self._ws = websocket.WebSocketApp(
            url,
            header=headers,
            on_open=self._on_open,
            on_message=self._on_message,
            on_error=self._on_error,
            on_close=self._on_close,
        )
        self._ws.run_forever(ping_interval=PING_INTERVAL, ping_timeout=10)

    def _on_open(self, ws):
        log.info("Connected ✓")

    def _on_message(self, ws, raw: str):
        try:
            msg = json.loads(raw)
        except json.JSONDecodeError:
            return

        msg_type = msg.get("type")
        if msg_type == "connected":
            log.info(f"Server: {msg.get('message')}")
        elif msg_type == "job":
            # 并发控制
            if not self._sem.acquire(blocking=False):
                log.warning(f"[Job #{msg.get('job_id')}] Worker busy, rejecting")
                self._send_result(ws, msg["job_id"], False, error="worker busy")
                return
            t = threading.Thread(target=self._handle_job, args=(ws, msg), daemon=True)
            t.start()
        elif msg_type == "ack":
            log.info(f"ACK: {msg.get('message')}")
        else:
            log.debug(f"Unknown message type: {msg_type}")

    def _on_error(self, ws, error):
        log.error(f"WS error: {error}")

    def _on_close(self, ws, code, reason):
        log.info(f"Disconnected (code={code} reason={reason})")

    def _handle_job(self, ws, msg: dict):
        try:
            job_id   = msg["job_id"]
            job_type = msg.get("job_type", "")
            queue    = msg.get("queue", "")
            payload  = msg.get("payload", "{}")
            tags     = msg.get("tags") or []   # v4: 任务标签

            log.info(f"[Job #{job_id}] type={job_type} queue={queue} tags={tags}")

            # 解析 payload
            try:
                data = json.loads(payload)
            except json.JSONDecodeError as e:
                self._send_result(ws, job_id, False, error=f"payload parse error: {e}")
                return

            # 查找 handler
            handler = self._handlers.get(job_type)
            if not handler:
                log.warning(f"[Job #{job_id}] No handler for job_type={job_type}")
                self._send_result(ws, job_id, False, error=f"no handler for job_type: {job_type}")
                return

            # 执行
            result = handler(data, tags)
            log.info(f"[Job #{job_id}] OK: {result}")
            self._send_result(ws, job_id, True, log_msg=result)

        except Exception as e:
            log.exception(f"[Job #{msg.get('job_id')}] FAILED: {e}")
            self._send_result(ws, msg.get("job_id"), False, error=str(e))
        finally:
            self._sem.release()

    def _send_result(self, ws, job_id: int, success: bool,
                     log_msg: str = "", error: str = ""):
        msg = {"type": "result", "job_id": job_id, "success": success}
        if success:
            msg["log"] = log_msg
        else:
            msg["error"] = error
        try:
            ws.send(json.dumps(msg))
        except Exception as e:
            log.error(f"Failed to send result: {e}")


# ─── 示例 Job Handlers ─────────────────────────────────────────────────────────

def handle_send_email(data: dict, tags: list) -> str:
    to      = data.get("to", "")
    subject = data.get("subject", "")
    log.info(f"[send_email] to={to} subject={subject} tags={tags}")
    time.sleep(0.3)
    return f"Email sent to {to}"


def handle_generate_report(data: dict, tags: list) -> str:
    name = data.get("name", "")
    log.info(f"[generate_report] name={name} tags={tags}")
    time.sleep(0.8)
    return f"Report '{name}' generated"


def handle_resize_image(data: dict, tags: list) -> str:
    url = data.get("url", "")
    w   = data.get("width", 0)
    h   = data.get("height", 0)
    log.info(f"[resize_image] url={url} size={w}x{h} tags={tags}")
    time.sleep(0.5)
    return f"Image {url} resized to {w}x{h}"


def handle_data_sync(data: dict, tags: list) -> str:
    src = data.get("source", "")
    dst = data.get("target", "")
    log.info(f"[data_sync] {src} → {dst} tags={tags}")
    time.sleep(0.6)
    return f"Synced {src} → {dst}"


def handle_tag_task(data: dict, tags: list) -> str:
    """v4: 演示如何根据 tags 做不同处理"""
    message = data.get("message", "")
    log.info(f"[tag_task] message={message} tags={tags}")

    if "dry-run" in tags:
        log.info("[tag_task] dry-run 模式，跳过实际操作")
        return "dry-run: skipped"
    if "urgent" in tags:
        log.info("[tag_task] 紧急任务，优先处理")
    if "notify" in tags:
        log.info("[tag_task] 需要发送通知")

    time.sleep(0.2)
    return f"tag_task done: {message} (tags={tags})"


def handle_batch_callback(data: dict, tags: list) -> str:
    """v4: 处理 batch 的 then/catch/finally 回调"""
    batch_id = data.get("batch_id", "")
    status   = data.get("status", "")
    log.info(f"[batch_callback] batch_id={batch_id} status={status} tags={tags}")
    # 在这里可以发送通知、更新数据库等
    return f"batch {batch_id} callback handled (status={status})"


# ─── 入口 ─────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    worker = GoQueueWorker()

    # 基础 handlers
    worker.register("send_email",      handle_send_email)
    worker.register("generate_report", handle_generate_report)
    worker.register("resize_image",    handle_resize_image)
    worker.register("data_sync",       handle_data_sync)

    # v4: tags 示例
    worker.register("tag_task",        handle_tag_task)

    # v4: batch 回调示例
    worker.register("batch_callback",  handle_batch_callback)
    worker.register("on_success",      handle_batch_callback)   # batch then_job
    worker.register("on_failure",      handle_batch_callback)   # batch catch_job
    worker.register("on_finally",      handle_batch_callback)   # batch finally_job

    log.info(f"GoQueue Python Worker starting... Queue={QUEUE} Concurrency={CONCURRENCY} (Ctrl+C to stop)")
    try:
        worker.run()
    except KeyboardInterrupt:
        worker.stop()
        log.info("Worker stopped.")
